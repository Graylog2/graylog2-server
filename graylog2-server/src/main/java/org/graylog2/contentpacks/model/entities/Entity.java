/**
 * This file is part of Graylog.
 *
 * Graylog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog2.contentpacks.model.entities;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import org.graylog2.contentpacks.model.Identified;
import org.graylog2.contentpacks.model.Typed;
import org.graylog2.contentpacks.model.Versioned;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = Versioned.FIELD_META_VERSION)
@JsonSubTypes({
        @JsonSubTypes.Type(value = EntityV1.class, name = EntityV1.VERSION)
})
public interface Entity extends Identified, Typed, Versioned {
    EntityDescriptor toEntityDescriptor();

    interface EntityBuilder<SELF> extends IdBuilder<SELF>, TypeBuilder<SELF>, VersionBuilder<SELF> {
    }
}
