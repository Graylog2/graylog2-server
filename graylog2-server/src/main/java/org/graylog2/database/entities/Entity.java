/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
package org.graylog2.database.entities;

import com.fasterxml.jackson.annotation.JsonProperty;
import org.mongojack.Id;
import org.mongojack.ObjectId;

import javax.annotation.Nullable;

/**
 * Entity base class, which can be used to enforce that each entity implementation
 * has the required id and metadata fields.
 */
// TODO: we might consider renaming these 'Entity' classes.  There is already the notion of 'Entities' in content pack management, which has an entirely different meaning.
public abstract class Entity {
    private static final String ID = "id";
    static final String METADATA = "_metadata";

    // The @Id and @ObjectId annotations take care of converting the "id" field name to "_id" when storing/loading
    // values to/from the database.
    @Id
    @ObjectId
    @Nullable
    @JsonProperty(ID)
    public abstract String id();

    @JsonProperty(METADATA)
    public abstract EntityMetadata metadata();

    public abstract <T extends Entity> T withMetadata(EntityMetadata metadata);

    public static abstract class Builder<SELF extends Builder<SELF>> {
        @Id
        @ObjectId
        @JsonProperty(ID)
        public abstract SELF id(@Nullable String id);

        @JsonProperty(METADATA)
        public abstract SELF metadata(EntityMetadata metadata);
    }
}
