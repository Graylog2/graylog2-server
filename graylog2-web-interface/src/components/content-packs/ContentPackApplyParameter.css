:local(.applyParameter) {
    margin-top: 8px;
}
