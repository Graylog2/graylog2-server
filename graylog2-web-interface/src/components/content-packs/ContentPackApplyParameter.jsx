import React from 'react';
import PropTypes from 'prop-types';
import { findIndex } from 'lodash';

import { Row, Col, Button } from 'react-bootstrap';
import { Input } from 'components/bootstrap';
import DataTable from 'components/common/DataTable';
import ValueReferenceData from 'util/ValueReferenceData';
import naturalSort from 'javascript-natural-sort';

import Style from './ContentPackApplyParameter.css';

class ContentPackApplyParameter extends React.Component {
  static propTypes = {
    onParameterApply: PropTypes.func,
    onParameterClear: PropTypes.func,
    entity: PropTypes.object.isRequired,
    parameters: PropTypes.array,
    appliedParameter: PropTypes.array,
  };

  static defaultProps = {
    onParameterApply: () => {},
    onParameterClear: () => {},
    parameters: [],
    appliedParameter: [],
  };

  constructor(props) {
    super(props);

    this.state = {
      config_key: '',
      parameter: '',
    };
  }

  _configKeyRowFormatter = (paramMap) => {
    const enableClear = findIndex(this.props.appliedParameter,
      { paramName: paramMap.paramName, configKey: paramMap.configKey, readOnly: true }) < 0;
    const lastCol = enableClear ?
      <td><Button bsStyle="info" bsSize="small" onClick={() => { this._parameterClear(paramMap.configKey); }}>Clear</Button></td> :
      <td />;
    return (
      <tr key={paramMap.configKey}>
        <td>{paramMap.configKey}</td>
        <td>{paramMap.paramName}</td>
        { lastCol }
      </tr>
    );
  };

  _bindValue = (event) => {
    const newValue = {};
    newValue[event.target.name] = event.target.value;
    this.setState(newValue);
  };

  _valuesSelected = () => {
    return this.state.parameter.length > 0 && this.state.config_key.length > 0;
  };

  _applyParameter = (e) => {
    e.preventDefault();
    if (!this._valuesSelected()) {
      return;
    }
    const configKeyIndex = this.props.appliedParameter.findIndex((appliedParameter) => {
      return appliedParameter.configKey === this.state.config_key;
    });
    if (configKeyIndex >= 0) {
      return;
    }
    this.props.onParameterApply(this.state.config_key, this.state.parameter);
    this.setState({ config_key: '', parameter: '' });
  };

  _parameterClear = (configKey) => {
    this.props.onParameterClear(configKey);
  };

  render() {
    const vRefData = new ValueReferenceData(this.props.entity.data);
    const configPaths = vRefData.getPaths();

    const configKeys = Object.keys(configPaths)
      .sort(naturalSort)
      .filter(configKey => configPaths[configKey].isValueRef()) // Only allow value-refs as parameters
      .filter((configKey) => {
        return this.props.appliedParameter.findIndex((paramMap) => {
          return configKey === paramMap.configKey;
        }) < 0;
      });
    const emptyOption = (name) => { return (<option key="EMPTY" value="">{name}</option>); };
    const configOptions = [emptyOption('Choose Config Key')].concat(configKeys.map(key => <option key={key} value={key}>{key}</option>));
    let parameters = this.props.parameters;
    let emptyName = parameters.length <= 0 ? 'Create a parameter first' : 'Choose...';
    if (this.state.config_key !== '' && parameters.length > 0) {
      const configKeyType = configPaths[this.state.config_key].getValueType();
      if (['string', 'integer', 'boolean', 'double'].findIndex(type => type === configKeyType) >= 0) {
        parameters = parameters.filter(parameter => parameter.type === configKeyType);
      }
      emptyName = parameters.length <= 0 ? `No parameter from type ${configKeyType}` : 'Choose...';
    }
    const parameterOptions = [emptyOption(emptyName)]
      .concat(parameters.map(key => <option key={key.name} value={key.name}>{key.title} ({key.name})</option>));
    return (
      <div>
        <form className="apply-parameter-form" id="apply-parameter-form" onSubmit={this._applyParameter}>
          <Row className={Style.applyParameter}>
            <Col smOffset={1} sm={5}>
              <Input name="config_key"
                     id="config_key"
                     type="select"
                     value={this.state.config_key}
                     onChange={this._bindValue}
                     label="Config Key"
                     required>
                {configOptions}
              </Input>
            </Col>
            <Col sm={5}>
              <Input name="parameter"
                     id="parameter"
                     type="select"
                     value={this.state.parameter}
                     onChange={this._bindValue}
                     label="Parameter"
                     required>
                {parameterOptions}
              </Input>
            </Col>
            <Col sm={1} />
          </Row>
          <Row>
            <Col smOffset={10} sm={2}>
              <Button bsStyle="primary" disabled={!this._valuesSelected()} type="submit">Apply</Button>
            </Col>
          </Row>
        </form>
        <Row>
          <Col smOffset={1} sm={10}>
            <DataTable
              id="config-key-list"
              headers={['Config Key', 'Parameter', 'Action']}
              filterKeys={[]}
              rows={this.props.appliedParameter}
              dataRowFormatter={this._configKeyRowFormatter}
            />
          </Col>
        </Row>
      </div>
    );
  }
}

export default ContentPackApplyParameter;
