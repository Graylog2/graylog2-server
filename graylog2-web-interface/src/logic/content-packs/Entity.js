import { Map } from 'immutable';
import { findIndex } from 'lodash';
import ValueRefHelper from 'util/ValueRefHelper';
import Constraint from './Constraint';

export default class Entity {
  constructor(v, type, id, data, fromServer = false, constraintValues = [], parameters = []) {
    const constraints = constraintValues.map((c) => {
      if (c instanceof Constraint) {
        return c;
      }
      return Constraint.fromJSON(c);
    });

    this._value = {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    };
  }

  static fromJSON(value, fromServer = true, parameters = []) {
    const { v, type, id, data, constraints } = value;
    return new Entity(v, type, id, data, fromServer, constraints, parameters);
  }

  get v() {
    return this._value.v;
  }

  get type() {
    return this._value.type;
  }

  get id() {
    return this._value.id;
  }

  get data() {
    return this._value.data;
  }

  get fromServer() {
    return this._value.fromServer;
  }

  get constraints() {
    return this._value.constraints;
  }

  get title() {
    let value = this.getValueFromData('title');
    if (!value) {
      value = this.getValueFromData('name');
    }
    return value || '';
  }

  get description() {
    return this.getValueFromData('description') || '';
  }

  /* eslint-disable-next-line class-methods-use-this */
  get isEntity() {
    return true;
  }

  /* implement custom instanceof */
  static [Symbol.hasInstance](obj) {
    if (obj.isEntity) {
      return true;
    }
    return false;
  }

  getValueFromData(key) {
    const { data } = this._value;
    if (!data || !data[key]) {
      return undefined;
    }

    if (ValueRefHelper.dataIsValueRef(data[key])) {
      const value = (data[key] || {})[ValueRefHelper.VALUE_REF_VALUE_FIELD];
      if (ValueRefHelper.dataValueIsParameter(data[key])) {
        const index = findIndex(this._value.parameters, { name: value });
        if (index >= 0 && this._value.parameters[index].default_value) {
          return this._value.parameters[index].default_value;
        }
      }
      return value;
    }
    return data[key];
  }

  toBuilder() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    } = this._value;
    /* eslint-disable-next-line no-use-before-define */
    return new Builder(Map({
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    }));
  }


  static builder() {
    /* eslint-disable-next-line no-use-before-define */
    return new Builder();
  }

  toJSON() {
    const {
      v,
      type,
      id,
      data,
      constraints,
    } = this._value;
    return {
      v,
      type,
      id,
      data,
      constraints,
    };
  }
}

class Builder {
  constructor(value = Map()) {
    this.value = value;
  }

  v(value) {
    this.value = this.value.set('v', value);
    return this;
  }

  type(value) {
    this.value = this.value.set('type', value);
    return this;
  }

  id(value) {
    this.value = this.value.set('id', value);
    return this;
  }

  data(value) {
    this.value = this.value.set('data', value);
    return this;
  }

  fromServer(value) {
    this.value = this.value.set('fromServer', value);
    return this;
  }

  constraints(value) {
    this.value = this.value.set('constraints', value);
    return this;
  }

  parameters(value) {
    this.value = this.value.set('parameters', value);
    return this;
  }

  build() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    } = this.value.toObject();
    return new Entity(v, type, id, data, fromServer, constraints, parameters);
  }
}
