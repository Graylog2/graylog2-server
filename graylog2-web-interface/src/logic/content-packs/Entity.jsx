import { Map } from 'immutable';
import ValueRefHelper from 'util/ValueRefHelper';
import Constraint from './Constraint';

export default class Entity {
  constructor(v, type, id, data, fromServer = false, constraintValues = []) {
    const constraints = constraintValues.map((c) => {
      if (c instanceof Constraint) {
        return c;
      }
      return Constraint.fromJSON(c);
    });

    this._value = {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
    };
  }

  static fromJSON(value, fromServer = true) {
    const { v, type, id, data, constraints } = value;
    return new Entity(v, type, id, data, fromServer, constraints);
  }

  get v() {
    return this._value.v;
  }

  get type() {
    return this._value.type;
  }

  get id() {
    return this._value.id;
  }

  get data() {
    return this._value.data;
  }

  get fromServer() {
    return this._value.fromServer;
  }

  get constraints() {
    return this._value.constraints;
  }

  get title() {
    let value = this.getValueFromData('title');
    if (!value) {
      value = this.getValueFromData('name');
    }
    return value || '';
  }

  get description() {
    return this.getValueFromData('description') || '';
  }

  getValueFromData(key) {
    const { data } = this._value;
    if (!data || !data[key]) {
      return undefined;
    }

    if (ValueRefHelper.dataIsValueRef(data[key])) {
      return (data[key] || {})[ValueRefHelper.VALUE_REF_VALUE_FIELD];
    }
    return data[key];
  }

  toBuilder() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
    } = this._value;
    /* eslint-disable-next-line no-use-before-define */
    return new Builder(Map({
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
    }));
  }


  static builder() {
    /* eslint-disable-next-line no-use-before-define */
    return new Builder();
  }

  toJSON() {
    const {
      v,
      type,
      id,
      data,
      constraints,
    } = this._value;
    return {
      v,
      type,
      id,
      data,
      constraints,
    };
  }
}

class Builder {
  constructor(value = Map()) {
    this.value = value;
  }

  v(value) {
    this.value = this.value.set('v', value);
    return this;
  }

  type(value) {
    this.value = this.value.set('type', value);
    return this;
  }

  id(value) {
    this.value = this.value.set('id', value);
    return this;
  }

  data(value) {
    this.value = this.value.set('data', value);
    return this;
  }

  fromServer(value) {
    this.value = this.value.set('fromServer', value);
    return this;
  }

  constraints(value) {
    this.value = this.value.set('constraints', value);
    return this;
  }

  build() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
    } = this.value.toObject();
    return new Entity(v, type, id, data, fromServer, constraints);
  }
}
