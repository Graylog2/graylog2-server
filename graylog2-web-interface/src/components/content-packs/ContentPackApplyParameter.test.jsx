import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import 'helpers/mocking/react-dom_mock';
import Entity from 'logic/content-packs/Entity';

import ContentPackApplyParameter from 'components/content-packs/ContentPackApplyParameter';

describe('<ContentPackApplyParameter />', () => {
  const entity = Entity.builder()
    .id('111-beef')
    .v('1.0')
    .data({
      name: { '@type': 'string', '@value': 'Input' },
      title: { '@type': 'string', '@value': 'A good input' },
      configuration: {
        listen_address: { '@type': 'string', '@value': '1.2.3.4' },
        port: { '@type': 'integer', '@value': '23' },
      },
    })
    .build();

  const parameter = { title: 'Port', name: 'PORT', type: 'integer', default_value: '23' };
  const appliedParameter = { configKey: 'configuration.port', paramName: parameter.name };
  const appliedParameterReadOnly = { configKey: 'configuration.port', paramName: parameter.name, readOnly: true };

  it('should render with full props', () => {
    const wrapper = renderer.create(<ContentPackApplyParameter
      entity={entity}
      parameters={[parameter]}
      appliedParameter={[appliedParameter]}
    />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with readOnly', () => {
    const wrapper = renderer.create(<ContentPackApplyParameter
      entity={entity}
      parameters={[parameter]}
      appliedParameter={[appliedParameterReadOnly]}
    />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with minimal props', () => {
    const wrapper = renderer.create(<ContentPackApplyParameter
      entity={entity}
    />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should apply a parameter', () => {
    const applyFn = jest.fn((configKey, paramName) => {
      expect(configKey).toEqual('configuration.port');
      expect(paramName).toEqual('PORT');
    });

    const wrapper = mount(<ContentPackApplyParameter
      entity={entity}
      parameters={[parameter]}
      appliedParameter={[]}
      onParameterApply={applyFn}
    />);

    wrapper.find('select#config_key').simulate('change', { target: { name: 'config_key', value: 'configuration.port' } });
    wrapper.find('select#parameter').simulate('change', { target: { name: 'parameter', value: 'PORT' } });
    wrapper.find('form').simulate('submit');
    expect(applyFn.mock.calls.length).toBe(1);
  });

  it('should apply a parameter only once', () => {
    const applyFn = jest.fn((configKey, paramName) => {
      expect(configKey).toEqual('configuration.port');
      expect(paramName).toEqual('PORT');
    });

    const wrapper = mount(<ContentPackApplyParameter
      entity={entity}
      parameters={[parameter]}
      appliedParameter={[{ configKey: 'configuration.port', paramName: 'PORT' }]}
      onParameterApply={applyFn}
    />);

    wrapper.find('select#config_key').simulate('change', { target: { name: 'config_key', value: 'configuration.port' } });
    wrapper.find('select#parameter').simulate('change', { target: { name: 'parameter', value: 'PORT' } });
    wrapper.find('form').simulate('submit');
    expect(applyFn.mock.calls.length).toBe(0);
  });

  it('should clear a parameter', () => {
    const clearFn = jest.fn((configKey) => {
      expect(configKey).toEqual('configuration.port');
    });

    const wrapper = mount(<ContentPackApplyParameter
      entity={entity}
      parameters={[parameter]}
      appliedParameter={[appliedParameter]}
      onParameterClear={clearFn}
    />);

    wrapper.find('button[children="Clear"]').simulate('click');
    expect(clearFn.mock.calls.length).toBe(1);
  });
});
