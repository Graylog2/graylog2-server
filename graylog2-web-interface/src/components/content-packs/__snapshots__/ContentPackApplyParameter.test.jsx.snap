// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentPackApplyParameter /> should render with full props 1`] = `
<div>
  <form
    className="apply-parameter-form"
    id="apply-parameter-form"
    onSubmit={[Function]}
  >
    <div
      className="applyParameter row"
    >
      <div
        className="col-sm-5 col-sm-offset-1"
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="config_key"
          >
            Config Key
          </label>
          <span>
            <select
              className="form-control"
              id="config_key"
              label="Config Key"
              name="config_key"
              onChange={[Function]}
              placeholder=""
              required={true}
              type="select"
              value=""
            >
              <option
                value=""
              >
                Choose Config Key
              </option>
              <option
                value="configuration.listen_address"
              >
                configuration.listen_address
              </option>
              <option
                value="name"
              >
                name
              </option>
              <option
                value="title"
              >
                title
              </option>
            </select>
            
          </span>
        </div>
      </div>
      <div
        className="col-sm-5"
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="parameter"
          >
            Parameter
          </label>
          <span>
            <select
              className="form-control"
              id="parameter"
              label="Parameter"
              name="parameter"
              onChange={[Function]}
              placeholder=""
              required={true}
              type="select"
              value=""
            >
              <option
                value=""
              >
                Choose...
              </option>
              <option
                value="PORT"
              >
                Port
                 (
                PORT
                )
              </option>
            </select>
            
          </span>
        </div>
      </div>
      <div
        className="col-sm-1"
      />
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-2 col-sm-offset-10"
      >
        <button
          className="btn btn-primary"
          disabled={true}
          type="submit"
        >
          Apply
        </button>
      </div>
    </div>
  </form>
  <div
    className="row"
  >
    <div
      className="col-sm-10 col-sm-offset-1"
    >
      <div>
        <div
          className="row "
        >
          <div
            className="col-md-12"
          >
            <div
              className="data-table table-responsive"
              id="config-key-list"
            >
              <table
                className="table undefined"
              >
                <thead>
                  <tr>
                    <th>
                      Config Key
                    </th>
                    <th>
                      Parameter
                    </th>
                    <th>
                      Action
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      configuration.port
                    </td>
                    <td>
                      PORT
                    </td>
                    <td>
                      <button
                        className="btn btn-sm btn-info"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Clear
                      </button>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ContentPackApplyParameter /> should render with minimal props 1`] = `
<div>
  <form
    className="apply-parameter-form"
    id="apply-parameter-form"
    onSubmit={[Function]}
  >
    <div
      className="applyParameter row"
    >
      <div
        className="col-sm-5 col-sm-offset-1"
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="config_key"
          >
            Config Key
          </label>
          <span>
            <select
              className="form-control"
              id="config_key"
              label="Config Key"
              name="config_key"
              onChange={[Function]}
              placeholder=""
              required={true}
              type="select"
              value=""
            >
              <option
                value=""
              >
                Choose Config Key
              </option>
              <option
                value="configuration.listen_address"
              >
                configuration.listen_address
              </option>
              <option
                value="configuration.port"
              >
                configuration.port
              </option>
              <option
                value="name"
              >
                name
              </option>
              <option
                value="title"
              >
                title
              </option>
            </select>
            
          </span>
        </div>
      </div>
      <div
        className="col-sm-5"
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="parameter"
          >
            Parameter
          </label>
          <span>
            <select
              className="form-control"
              id="parameter"
              label="Parameter"
              name="parameter"
              onChange={[Function]}
              placeholder=""
              required={true}
              type="select"
              value=""
            >
              <option
                value=""
              >
                Create a parameter first
              </option>
            </select>
            
          </span>
        </div>
      </div>
      <div
        className="col-sm-1"
      />
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-2 col-sm-offset-10"
      >
        <button
          className="btn btn-primary"
          disabled={true}
          type="submit"
        >
          Apply
        </button>
      </div>
    </div>
  </form>
  <div
    className="row"
  >
    <div
      className="col-sm-10 col-sm-offset-1"
    >
      <div>
        <div
          className="row "
        >
          <div
            className="col-md-12"
          >
            <div
              className="data-table table-responsive"
              id="config-key-list"
            >
              <p>
                No data available.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ContentPackApplyParameter /> should render with readOnly 1`] = `
<div>
  <form
    className="apply-parameter-form"
    id="apply-parameter-form"
    onSubmit={[Function]}
  >
    <div
      className="applyParameter row"
    >
      <div
        className="col-sm-5 col-sm-offset-1"
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="config_key"
          >
            Config Key
          </label>
          <span>
            <select
              className="form-control"
              id="config_key"
              label="Config Key"
              name="config_key"
              onChange={[Function]}
              placeholder=""
              required={true}
              type="select"
              value=""
            >
              <option
                value=""
              >
                Choose Config Key
              </option>
              <option
                value="configuration.listen_address"
              >
                configuration.listen_address
              </option>
              <option
                value="name"
              >
                name
              </option>
              <option
                value="title"
              >
                title
              </option>
            </select>
            
          </span>
        </div>
      </div>
      <div
        className="col-sm-5"
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="parameter"
          >
            Parameter
          </label>
          <span>
            <select
              className="form-control"
              id="parameter"
              label="Parameter"
              name="parameter"
              onChange={[Function]}
              placeholder=""
              required={true}
              type="select"
              value=""
            >
              <option
                value=""
              >
                Choose...
              </option>
              <option
                value="PORT"
              >
                Port
                 (
                PORT
                )
              </option>
            </select>
            
          </span>
        </div>
      </div>
      <div
        className="col-sm-1"
      />
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-2 col-sm-offset-10"
      >
        <button
          className="btn btn-primary"
          disabled={true}
          type="submit"
        >
          Apply
        </button>
      </div>
    </div>
  </form>
  <div
    className="row"
  >
    <div
      className="col-sm-10 col-sm-offset-1"
    >
      <div>
        <div
          className="row "
        >
          <div
            className="col-md-12"
          >
            <div
              className="data-table table-responsive"
              id="config-key-list"
            >
              <table
                className="table undefined"
              >
                <thead>
                  <tr>
                    <th>
                      Config Key
                    </th>
                    <th>
                      Parameter
                    </th>
                    <th>
                      Action
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      configuration.port
                    </td>
                    <td>
                      PORT
                    </td>
                    <td />
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
